internal/executor/build/build.go

```diff
@@ -0,0 +1,101 @@
+package build
+
+import (
+	"github.com/cirruslabs/cirrus-ci-agent/api"
+	"github.com/cirruslabs/cirrus-cli/internal/executor/instance"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"path/filepath"
+	"sync"
+)
+
+type Build struct {
+	// A directory on host where .cirrus.yml that drives this execution is located
+	ProjectDir string
+
+	// Environment variables that will be injected via an agent
+	Environment map[string]string
+
+	// The actual tasks comprising this build
+	tasks map[int64]*Task
+
+	// A mutex to guarantee safe access to this build from both the "main loop" and gRPC server handlers
+	Mutex sync.Mutex
```

Comment:
I haven't seen any build access pattern that was not thread safe. I think it will be safe to remove Mutex. Did I miss something? Once we have parallel tasks and a fancy cmd UI then we'll need some synchronization there.

Reply:
I don't like this either, but this is not just about the locking per se, more about the memory barriers and visibility.

Without following the intricate rules of the [Go memory model](https://golang.org/ref/mem) (which is not recommended, see "Advice" section) or implementing an explicit synchronization with e.g. `sync.Mutex`, there's no guarantee that we'll see up-to-date task status values in `Executor.Run()`, since gRPC server that we also run handles requests and updates status values [in separate goroutines](https://stackoverflow.com/questions/50365652/does-the-grpc-server-use-one-new-goroutine-to-process-one-method-call), which easily translates to "runs in a separate thread", which in turn [can easily cause problems](https://en.wikipedia.org/wiki/Memory_barrier#Example).

Reply:
Yes, task's status is the only thing that changes and CLI need to have updated value for figuring out which task to run next. We can just use some atomic in order to do the updates (make status private and use getter and setter for the atomic operation). What do you think?

Reply:
I thought about wrapping it in an interface method too, but don't you think that in the near future more fields will show up, and it would basically end up with the same mutex wraps?

Reply:
I just don't like concurrency primitives except channels and atomics. In most of the cases use of a Mutex overcomplicates things. I'll think how we can re-architecture things like `instance`, `executor` and `build`. 🤔

Reply:
On the second though it seems we'll have serialization of the status update by nature of how things are executed at the moment: a status is updated **before container is killed** and the access to the task status happens **after container is cleaned up** in order to start a new task.

I'm currently also don't like how `instance` is responsible for executing itself. It simplifies things when we are running one task at a time but with the parallel execution we'll need to figure out some sort of instance manager that will be aware of local resources available to the CLI and run as many tasks in parallel as possible and there will need synchronization. Probably with some sort of a channel with task updates.

Reply:
> On the second though it seems we'll have serialization of the status update by nature of how things are executed at the moment: a status is updated **before container is killed** and the access to the task status happens **after container is cleaned up** in order to start a new task.

You're right, but The Go Memory Model document clearly [recommends against this](https://golang.org/ref/mem#tmp_1):

>Programs that modify data being simultaneously accessed by multiple goroutines must serialize such access.
>
>To serialize access, protect the data with channel operations or other synchronization primitives such as those in the sync and sync/atomic packages.
>
>If you must read the rest of this document to understand the behavior of your program, you are being too clever.
>
>Don't be clever.

I'll implement the getter/setter methods as discussed above, and if you will keep the history while merging this, we will be able to reference the `sync.Mutex` solution any time in the future (props to Git).

>I'm currently also don't like how `instance` is responsible for executing itself.

Is it? As far as I can see, it is being executed by the `Executor` by calling the `Run()` method on an `Instance`.

>It simplifies things when we are running one task at a time but with the parallel execution we'll need to figure out some sort of instance manager that will be aware of local resources available to the CLI and run as many tasks in parallel as possible and there will need synchronization. Probably with some sort of a channel with task updates.

I agree that an instance manager makes sense in the future, but hey, this is a basic implementation just so we can start moving forward!

Reply:
Ok, so we are agreeing on moving the mutex to task and having getter / setter? I'll be OK with that and we can revisit it once we'll think about the parallel execution.

Reply:
Added `sync.Mutex`-wrapped getter and setter for the task's state field in https://github.com/cirruslabs/cirrus-cli/pull/6/commits/a9a2096067d4dab3ef6fb866d95ea0cc465bb3ba.